addons/mrp/models/product.py

```diff
@@ -28,7 +28,7 @@ def _compute_bom_count(self):
     def _compute_used_in_bom_count(self):
         for template in self:
             template.used_in_bom_count = self.env['mrp.bom'].search_count(
-                [('bom_line_ids.product_id', 'in', template.product_variant_ids.ids)])
+                [('bom_line_ids.product_id.product_tmpl_id', '=', template.id)])
```

Comment:
Nice one. Do you think it could be even faster with only one read_group before the loop ?

Reply:
it's used in form views only, so I don't think that further optimization is really needed

Reply:
You can simplify it because of the related in mrp.bom.line (should change the SQL query behind):
`[('bom_line_ids.product_tmpl_id', '=', template.id)])`.

Also, a alternative solution is to index the `product_id` of the `mrp.bom.line` (drawback is it needs a upgrade) and keep the old version of the compute. I tested, it should be faster but I don't know it is worth it: In your version, how much take `_compute_used_in_bom_count` on the 1.5 of the total request ? What do you think about the index solution ?

Reply:
indexing should speed it up. Ideal solution is searching by product_tmpl_id + indexing product_tmpl_id.

> how much take _compute_used_in_bom_count on the 1.5 of the total request 

only query:
before: 9,2 sec
after: 0,7 sec


About shorter domain by using related field. I'd prefer explicit version :-)

Reply:
Well agree, unfortunately, we can't index the `product_tmpl_id` because it is no store related (maybe change that in master, if it is worth it).
IMO, this fix is perfect for stable (still a huge improvement and doesn't need a upgrade) but for master I suppose that it will more efficient with the index of the product_id (0.7 sec, it seems still a lot just for show "used in") and keep the compute in the previous version. Can you test in index on your DB to see the result (and support my previous test I hope) ?

"About shorter domain by using related field. I'd prefer explicit version :-)" -> Yes it seems legit

Reply:
After adding ``store=True, index=True`` + using domain ``[('bom_line_ids.product_tmpl_id', '=', template.id)])`` execution time drops to 1ms. Maybe add it to this PR? It will work even without upgrading the module and should not break anything if module is  not upgraded

Reply:
Well, actually it's not safe. PR to master: https://github.com/odoo/odoo/pull/69293

Reply:
> You can simplify it because of the related in mrp.bom.line (should change the SQL query behind):
> [('bom_line_ids.product_tmpl_id', '=', template.id)]).

Done to avoid conflicts with master PR

Reply:
IMO:
 - In stable version: you can't store `product_tmpl_id` in stable. Then either take your new compute (no need upgrade) or take add index on `product_id` (need a upgrade :( )
 - In master: either only add index on `product_id` (simple) or add store=True, index=True on `product_tmpl_id` (but add duplication in DB)

Reply:
You change here but not for the domain in action `action_used_in_bom`.

Reply:
> You change here but not for the domain in action `action_used_in_bom`.

done